riscv/triggers.cc

```diff
@@ -61,7 +61,7 @@ bool trigger_t::common_match(processor_t * const proc) const noexcept {
 
 bool trigger_t::mode_match(state_t * const state) const noexcept
 {
-  switch (state->prv) {
+  switch (state->last_inst_priv) {
```

Comment:
This change will apply to `icount` and `mcontrol` type breakpoints as well, and will break them, because it will be considering the privilege of the previous instruction instead of the current.

It seems that it's only for `detect_trap_match()` that we want this behavior.

Could you get that behavior by calling `detect_trap_match()` just before calling `take_trap()`? I suspect that would work, and then you wouldn't need to use `last_inst_priv` at all.

Reply:
@scottj97 A coworker suggested something along the lines of the following based on your comment. It seems reasonable, but maybe there be dragons?

```
diff --git a/riscv/execute.cc b/riscv/execute.cc
index 295879d4..b92fffdb 100644
--- a/riscv/execute.cc
+++ b/riscv/execute.cc
@@ -304,16 +304,19 @@ void processor_t::step(size_t n)
     }
     catch(trap_t& t)
     {
-      take_trap(t, pc);
       n = instret;
 
       // Trigger action takes priority over single step
       auto match = TM.detect_trap_match(t);
-      if (match.has_value())
+      if (match.has_value()) {
+        take_trap(t, pc);
         take_trigger_action(match->action, 0, state.pc, 0);
-      else if (unlikely(state.single_step == state.STEP_STEPPED)) {
-        state.single_step = state.STEP_NONE;
-        enter_debug_mode(DCSR_CAUSE_STEP);
+      } else {
+        take_trap(t, pc);
+        if (unlikely(state.single_step == state.STEP_STEPPED)) {
+          state.single_step = state.STEP_NONE;
+          enter_debug_mode(DCSR_CAUSE_STEP);
+        }
       }
     }
     catch (triggers::matched_t& t)
```

Reply:
In `detect_trap_match()` you need to know both the pre-trap privilege (for `common_match()`) and post-trap privilege (for `allow_action()`), yes? I could be wrong about this but it seems so since the breakpoint trap is being taken from the post-trap privilege (mepc will point to the first instruction of the handler, yes?).

I don't see how this proposed change could possibly do that.

Reply:
Also if it turns out that I'm wrong and your proposed change above is correct, I think it could be done much cleaner by simply moving the `take_trap()` call to be between the `auto match` and `if match` lines.

Reply:
@scottj97 Since it's a little difficult to talk about things in the abstract, let's consider the following scenario:

1. Trigger is set on an ECALL from U-mode
2. U-mode ECALL instruction is at pc = 0xAA
3. MTVEC handler at pc = 0xBB

It seems that with the suggested change, when pc = 0xAA:

- `detect_trap_match()` will succeed because we still haven't changed the privilege level. This is effectively the "pre-trap privilege level"
- Following the match, `take_trap()` is called. This will result in the "post-trap" privilege level (M) and  state.pc = 0xBB, which is the first instruction of the MTVEC handler as you noted. The mepc is set to 0xAA
-` take_trigger_action()` is now called with the new pc (0xBB)

In my mind, this is indeed the expected behavior. As you pointed out, we can implement it more succinctly by moving` trap_trap()` after the `auto match`, but it seems like the approach should work.

Reply:
Assume that neither ECALL nor breakpoint traps are delegated beyond M-mode. Assume an etrigger with action=0 (breakpoint exception).

The problem in your example is that the trigger action should have been suppressed by the mechanism in #1316, which is intended to prevent a reentrant M-mode exception from clobbering `mepc` and making recovery impossible. Inside `detect_trap_match()`, it calls `allow_action()` which should detect that we're already in M-mode and `mstatus.mie==0` therefore we should not allow this action.

It will fail to detect this because privilege is still set to U-mode. The result will be two consecutive M-mode exceptions, and runaway software.

Reply:
@scottj97 Thanks for pointing out the issue with the alternative approach.

Based on your previous feedback, it seems that one of the main issues with matching on the previous privilege was that it was causing icount match tests to fail. Having looked at the code again, the likely cause was that all those code paths use  `common_match()`.  I have attempted to resolve that issue by special casing the trigger matches for traps...PTAL.